internal/controller/logger_manager.go

```diff
@@ -0,0 +1,267 @@
+
+	// If previously logs were lost, send a log message to indicate that
+	if b.lostLogs {
+		lostLogMessage := zapcore.Entry{
+			Level:   zapcore.ErrorLevel,
+			Time:    time.Now(),
+			Message: b.lostLogMessage(),
+		}
+		err := b.sendLog(lostLogMessage)
+		if err != nil {
+			b.logger.Warn("Error while sending log",
+				"error", err,
+			)
+		}
+		b.lostLogs = false
+		b.lostBytes = 0
+	}
+
+}
+
+// run flushes the buffer at the configured interval until Stop is called.
+func (b *BufferedGrpcWriteSyncer) run() {
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			b.flush()
+		case <-b.done:
+			return
+		}
+	}
+}
+
+// sendLog sends the log as a string to the server.
+func (b *BufferedGrpcWriteSyncer) sendLog(logEntry zapcore.Entry) error {
+	buf, err := b.encoder.EncodeEntry(logEntry, nil)
+	if err != nil {
+		return err
+	}
+	err = b.client.Send(&pb.SendLogsRequest{
+		Request: &pb.SendLogsRequest_LogEntry{
+			LogEntry: &pb.LogEntry{
+				Level:       pb.LogLevel(logEntry.Level),
+				Time:        timestamppb.New(logEntry.Time),
+				JsonMessage: buf.String(),
+			},
+		},
+	})
+	return err
+}
+
+// UpdateClient will update BufferedGrpcWriteSyncer with new client stream and GRPC connection
+func (b *BufferedGrpcWriteSyncer) UpdateClient(client pb.KubernetesInfoService_SendLogsClient, conn *grpc.ClientConn) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	b.client = client
+	b.conn = conn
+
+	// If logs were lost, send a log message about it
+	if b.conn.GetState() == connectivity.Ready && b.lostLogs {
+		lostLogMessage := zapcore.Entry{
+			Level:   zapcore.ErrorLevel,
+			Time:    time.Now(),
+			Message: b.lostLogMessage(),
+		}
+		err := b.sendLog(lostLogMessage)
+		if err != nil {
+			b.logger.Warn("Error while sending log",
+				"error", err,
+			)
+		}
+		b.lostLogs = false
+		b.lostBytes = 0
+	}
+}
+
+// ListenToLogStream will wait for responses from server and will update log level
+// depending respone contents
+func (b *BufferedGrpcWriteSyncer) ListenToLogStream() {
+	for {
+		res, err := b.client.Recv()
+		if err == io.EOF {
+			// The client has closed the stream
+			b.logger.Info("Server has closed the SendLogs stream")
+			return
+		}
+		if err != nil {
+			b.logger.Errorw("Stream terminated",
+				"error", err,
+			)
+			return
+		}
+		switch res.Response.(type) {
+		case *pb.SendLogsResponse_SetLogLevel:
+			newLevel := res.GetSetLogLevel().Level
+			b.updateLogLevel(newLevel)
+		}
+	}
+}
+
+// bufferLog adds the log entry to in-memory buffer
+func (b *BufferedGrpcWriteSyncer) bufferLog(entry zapcore.Entry) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	b.buffer = append(b.buffer, entry)
+	if len(b.buffer) >= maxBufferSize {
+		b.flush()
+	}
+}
+
+// lostLogMessage generates a message about lost logs
+func (b *BufferedGrpcWriteSyncer) lostLogMessage() string {
+	return "Lost logs due to buffer overflow. Total lost bytes: " + strconv.Itoa(b.lostBytes)
+}
```

Comment:
No need for a separate method for this. Just inline this.
And pass the lost count as a Zap field. Keep logs structured and avoid doing string formatting in log messages.